### ISSUE TITLE
Conch SFTP client fails at the end of a file listing

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3009]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3009|
|Type|defect|
|Created|2008-01-23 15:22:28Z|
|Branch|https://github.com/twisted/twisted/tree/sftp-listing-eof-3009|

```
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/filetransfer.py", line 52, in dataReceived
    f(data)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/filetransfer.py", line 694, in packet_STATUS
    msg, data = getNS(data)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/common.py", line 55, in getNS
    l, = struct.unpack('!L',s[c:c+4])
struct.error: unpack str size does not match format
```

This appears to be due to incorrect ```FX_EOF``` handling.  In ```packet_STATUS``` in this case, ```code == FX_EOF``` and ```data == ''```.  The method tries to decode the data anyway (with ```getNS```) which fails.  Later in the method, there's a check to handle ```FX_EOF```, but execution never gets there.

Attachments:

* [sftp-status-bug.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/269/269f4f812dd5574240101fe2585d2e5b9ac8ba2c/dccf3b339b561fe664d38a5bbcc49b2315712458.patch) (647 bytes) - added by exarkun on 2010-05-16 16:32:22Z - 
* [sshbug.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/269/269f4f812dd5574240101fe2585d2e5b9ac8ba2c/2c064392e7ef0fb3efc4b610b7fe4deaa700c9eb.py) (4763 bytes) - added by exarkun on 2010-05-16 17:19:02Z - example which produces the error
* [short-status-packet.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/269/269f4f812dd5574240101fe2585d2e5b9ac8ba2c/12f44e6bc61454f3d5148df9e4a7d93b426569ab.patch) (2168 bytes) - added by jcollie on 2011-08-29 20:39:43Z - Revised patch with tests
* [short-status-packet-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/269/269f4f812dd5574240101fe2585d2e5b9ac8ba2c/5f82ad9fdd6b55c3fdc86e651f4893f318876433.patch) (2691 bytes) - added by jcollie on 2011-09-06 18:27:37Z - re-revised patch with tests

<details><summary>Searchable metadata</summary>

```
trac-id__3009 3009
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_sftp_listing_eof_3009 branches/sftp-listing-eof-3009
branch_author__mt__jcollie mt, jcollie
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1201101748000000 1201101748000000
changetime__1315590512000000 1315590512000000
version__None None
owner__exarkun exarkun
cc__exarkun cc__mt
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="KDomke's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|KDomke commented|
|-|-|

Hi,

Here is the code snipped which broke because of the error:

http://groups.google.de/group/comp.lang.python/browse_thread/thread/dd922671819059a2/8f68b080705d65cb

Hope, that helps...

Kristian

**Comment by twisted-trac** [comment]
|<img alt="mt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|mt commented|
|-|-|

Hi guys,

I just ran into this issue when using filetransfer as an sftp client.  The specific issue I'm hitting is:

1) Establish an sftp connection over ssh channel.

2) Issue an openFile request.

3) In the openFile callback, issue a writeChunk request.

4) Per the sftp ietf draft (http://tools.ietf.org/html/draft-ietf-secsh-filexfer-13) section 8.2.3 on a file write:

  "The server responds to a write request with a SSH_FXP_STATUS message."

5) dataReceived is then invoked, which invokes packet_STATUS.  Packet status does a _parseRequest to get the deferred to callback on, and then proceeds to parse out the response.

6) First the response code is parsed out, which is FX_OK.

7) Next the msg, and lang are parsed out.  THIS IS WHERE the error occurs.

Now per the ietf doc, section 9.1, Status response returns the following structure:

       byte   SSH_FXP_STATUS
       uint32 request-id
       uint32 error/status code
       string error message (ISO-10646 UTF-8 [RFC-2279])
       string language tag (as defined in [RFC-1766])
       error-specific data

The issue at hand appears to be that the error message and language tag are ONLY included when an error status is returned.  So, in the case of an FX_OK, they are blank.

However, the code in packet_STATUS is as follows:

    def packet_STATUS(self, data):
        d, data = self._parseRequest(data)
        code, = struct.unpack('!L', data[:4])
        data = data[4:]
        msg, data = getNS(data)
        lang = getNS(data)
        if code == FX_OK:
            d.callback((msg, lang))
        elif code == FX_EOF:
            d.errback(EOFError(msg))
        elif code == FX_OP_UNSUPPORTED:
            d.errback(NotImplementedError(msg))
        else:
            d.errback(SFTPError(code, msg, lang))

Notice that it attempts to parse out the error msg and language tag even on an FX_OK.

The fix I made in my local copy is as follows:

  def packet_STATUS(self, data):
        d, data = self._parseRequest(data)
        code, = struct.unpack('!L', data[:4])

        if code == FX_OK:
            d.callback(code)
        else:
            data = data[4:]
            msg, data = getNS(data)
            lang = getNS(data)
            if code == FX_EOF:
                d.errback(EOFError(msg))
            elif code == FX_OP_UNSUPPORTED:
                d.errback(NotImplementedError(msg))
            else:
                d.errback(SFTPError(code, msg, lang))

I'm not sure if this is correct or not (ie, returning the code with the callback), but it resolves my issue (and I can't think of anything else to return for the result, perhaps an empty string?)

I ran a global unit test via "trial twisted" with my fix in place, and the result was:

(skips=393, expectedFailures=11, successes=5239)

So, it doesn't appear to have broken anything.  I'll read through patch process and attempt to submit a patch for this sometime next week (and try and add a unit test also), but I thought I'd add a comment in the mean time.

- MT

**Comment by twisted-trac** [comment]
|<img alt="mt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|mt commented|
|-|-|

An update, after playing around with filetransfer for a while, I ended up with the following solution:

   def packet_STATUS(self, data):
        d, data = self._parseRequest(data)
        code, = struct.unpack('!L', data[:4])
        data = data[4:]
        if data:
            msg, data = getNS(data)
            lang = getNS(data)
        else:
            msg, lang = '', ''
        if code == FX_OK:
            d.callback((msg, lang))
        elif code == FX_EOF:
            d.errback(EOFError(msg))
        elif code == FX_OP_UNSUPPORTED:
            d.errback(NotImplementedError(msg))
        else:
            d.errback(SFTPError(code, msg, lang))

This is because some errors (FX_FAILURE) don't return the msg/lang.

- MT

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Thanks for looking in to this.  Can you attach the changes as a unified diff instead?  Those tend to keep better.  And if you can write the test you mentioned, that'd be great also. :)  When you think the ticket is ready for a committer to consider applying it, just add the "review" keyword to be sure someone notices.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I attached the original example and the proposed fix as a diff.  Mostly what's left to do is turn the example into a unit test.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="jcollie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|jcollie commented|
|-|-|

OK, I've attached a new patch WITH TESTS for this bug.  It's a little bit different from the original patch in that it returns a zero-length string instead of None.  The original patch also has a bug in that lang would be a tuple because getNS returns a tuple.  The tests are pretty basic at this point, there's lots more room for expansion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to jcollie|
|-|-|

Thanks!  This basically looks good, I just have a few comments:

  1. The test methods should return the Deferred that they add a callback to.  Otherwise it's not clear that the callback actually gets run.  Trial will consider the test complete as soon as the test method returns, whereas if the Deferred is returned it will only consider the test complete when the Deferred fires, and it will fail the test if the Deferred fires with a failure.
  1. The behavior being verified by each test method should be documented in docstrings.
  1. New tests should use the `test_packetSTATUSShort` convention rather than the older `testPacketSTATUSShort` convention.
  1. It's probably worth adding a direct reference to the part of the RFC that defines this part of the protocol.  I think this is section 9.1 of <http://tools.ietf.org/id/draft-ietf-secsh-filexfer-13.txt>, where I do not see any indication that the message and language are optional, so it may also be worth noting that these extra checks are for interop with other non-compliant implementations.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="jcollie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|jcollie commented|
|-|-|

OK, I revised the patch according to your suggestions.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [32520]) Branching to 'sftp-listing-eof-3009'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32521]) Apply short-status-packet-2.patch

refs [#3009](https://github.com/twisted/twisted/issues/3009)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [32522]) Add one more unit test to cover the last branch; also some docstrings and a news fragment.

refs [#3009](https://github.com/twisted/twisted/issues/3009)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks again jcollie.  I added one more test for the case where msg is present but lang is not, since the code was also handling that.  I also made a few other minor coding style tweaks, and added a news fragment for the change (visible in the revision linked above, if you're interested).  Otherwise the patch looks great, so I'll go ahead and apply it.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [32523]) Merge sftp-listing-eof-3009

Author: mt, jcollie
Reviewer: exarkun
Fixes: [#3009](https://github.com/twisted/twisted/issues/3009)

Change `twisted.conch.ssh.filetransfer` to handle short `STATUS` packets,
rather than raising an unhandled exception while trying to parse them.  This
improves interoperability with other SFTP implementations which may leave the
message and/or language values out of the packet.
